module load r/4.3.0

1. Differential DNA methylation analysis

library(methylKit)
library(data.table)
library(GenomicRanges)
library(genomation)
library(ggplot2)
library(edmr)
library(IRanges)
library(mixtools)

file.list <- list("/path/to/file1_R1_001_trimmed_bismark_hisat2.sam.CpG_report.txt", "/path/to/file2_R1_001_trimmed_bismark_hisat2.sam.CpG_report.txt", "/path/to/file3_R1_001_trimmed_bismark_hisat2.sam.CpG_report.txt", ...) ### These are the methylation calls generated by Bismark.

mydata <- methRead(file.list, sample.id = list("file1", "file2", "file3", ...), assembly = "canetoad_CpG", header = FALSE, resolution = "base", pipeline = "bismarkCytosineReport", treatment = c(0, 0, 0, 0, ..., 1, 1, 1, ...), context = "CpG", mincov = 10)

filtered.data <- filterByCoverage(mydata, lo.count = 10, lo.perc = NULL, hi.count = NULL, hi.perc = 99.9, num.cores=2)

normalised.reads <- normalizeCoverage(filtered.data)

mydata_merge <- unite(normalised.reads, destrand = TRUE, min.per.group = 7L) ### Need to adjust min.per.group accordingly. Guideline = 1/2 group sample size + 1.

getCorrelation(mydata_merge, plot=TRUE)

clusterSamples(mydata_merge, dist = "correlation", method = "ward.D", plot = TRUE)

mydata_myDiff <- calculateDiffMeth(mydata_merge, adjust = "SLIM")

myDiff <- getMethylDiff(mydata_myDiff, difference = 20, qvalue = 0.05) ### Generates file with differentially methylated CpGs. Might want to adjust methylation difference.

write.table(myDiff, file = "/path/to/DMC.txt", sep = "\t", row.names = FALSE, col.names= FALSE, quote = FALSE)

mydata_dmr <- edmr(mydata_myDiff, DMC.methdiff = 20, DMR.methdiff = 10, mode = 1, ACF = TRUE)

mydata_DMRs <- as.data.frame(mydata_dmr)

mydata_dmr.filtered <- filter.dmr(mydata_dmr, DMR.qvalue = 0.05, mean.meth.diff = 20, num.CpGs = 5, num.DMCs = 3) ### Generates file with differentially methylated regions. Might want to adjust parameters here too.

mydata_dmr.filtered <- as.data.frame(mydata_dmr.filtered)

write.table(mydata_dmr.filtered, file = "/path/to/DMR.txt", sep = "\t", row.names = FALSE, col.names= FALSE, quote = FALSE)

2.
